### PR TITLE
 Fix ICE in rustdoc when merging generic and where bounds of an Fn with an output

### LINKED ISSUE #57180: Rustdoc ICEs when a dependency has duplicate generic bounds to functions returning tuples
Minimized from a failing rustc PR updating polonius, itself coming with a newer revision of datafrog.

There's a minimized reproduction [here](https://github.com/lqd/docfail).
If you add this crate as a git dependency, and use it, like so:
```rust
use docfail::Leaper;

fn main() {
}
```

Rustdoc will crash while trying to build the doc for this dependent crate (note that it does not fail building docs in the `docfail` repo itself)

Crash details:
```
Documenting repro v0.1.0 (/Users/rrakic/Documents/rust/docfail)
thread '<unnamed>' panicked at 'assertion failed: output.is_none()', src/librustdoc/clean/simplify.rs:108:21
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.32.0-nightly (14997d56a 2018-12-05) running on x86_64-apple-darwin

error: Could not document `repro`.

```

The essence of the problem is the duplicate bounds on an impl inside the dependency:
```rust
pub trait Leaper {}

pub struct FilterAnti<Func>
{
    phantom: ::std::marker::PhantomData<Func>,
}

impl<Func: Fn() -> (u32)>
    Leaper for FilterAnti<Func>
where
    Func: Fn() -> (u32),
{
}
```

Removing either of the duplicate `Fn(...)` bound will work around the crash. The bound requires that the closure return a tuple, and the rustdoc trace seems to be code related to parentheses.

I can reproduce the crash on stable 1.31.0 `(abe02cefd 2018-12-04)`, 1.31.1 `(b6c32da9b 2018-12-18)`, an older nightly from the beginning of December `(14997d56a 2018-12-05)`, but haven't bisected older releases.

I couldn't find other issues mentioning this, and the error only appeared in the aforementioned rust PR.

Thanks to @Mark-Simulacrum for the help in locating this quite uncommon issue.

### PR DESCRIPTION
**Pull request by Nashenas88**
Fixes #57180

### CODE REVIEW

**Comment by JohnCSimon** [comment]
Ping from triage
@QuietMisdreavus can you please review this?
Thank you.

**Comment by JohnCSimon** [comment]
Ping from triage
@QuietMisdreavus @Nashenas88 This PR hasn't been updated in the last six days. Can you post any status?
Thank you.

**Comment by Nashenas88** [comment]
I was doing some analysis on the originating cause (documented in #57180). I mostly believe this is still the proper fix, but I was hoping someone familiar with this codebase could review and help bump up the confidence here. I requested @QuietMisdreavus for review since she was helping me track down the bug during impl days.

**Comment by Alexendoo** [comment]
Ping from triage: requesting a review from @rust-lang/rustdoc

File: src/test/rustdoc/auxiliary/issue-57180.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+pub struct Struct<F>
+{
+    _p: ::std::marker::PhantomData<F>,
+}
+
+impl<F: Fn() -> u32>
+Trait for Struct<F>
+    where
+        F: Fn() -> u32,
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/test/rustdoc/auxiliary/issue-57180.rs]: Existence alone is fine.

**Comment by GuillaumeGomez** [comment]
Since the real change is only one line, pretty quick to review. The regression test is also present so all good. Don't hesitate to ping me or @rust-lang/rustdoc directly next time.

Thanks!

@bors: r+

**Pull request merged**